cirq-core/cirq/transformers/target_gatesets/approximate_two_qubit_gateset.py

```diff
@@ -0,0 +1,124 @@
+            sample_scaling: Relative number of random gate products to use in the
+                tabulation. The total number of random local unitaries scales as
+                ~ $max_infidelity^{-3/2} * sample_scaling$. Must be positive.
+            allow_missed_points: If True, the tabulation is allowed to conclude
+                even if not all points in the Weyl chamber are expected to be
+                compilable using 2 or 3 base gates. Otherwise an error is raised
+                in this case.
+            random_state: Random state or random state seed.
+
+        Raises:
+            ValueError: if base_gate is not a two qubit gate.
+        """
+        if base_gate.num_qubits() != 2:
+            raise ValueError(
+                "base_gate requires a two qubit gate. Given"
+                f" {str(base_gate)} which is {base_gate.num_qubits()} qubits."
+            )
+
+        super().__init__(
+            base_gate,
+            ops.MeasurementGate,
+            ops.AnyUnitaryGateFamily(1),
+            name=f'Approximate{str(base_gate)}Gateset.',
+        )
+        self._base_gate = base_gate
+        self._tabulation = two_qubit_gate_tabulation.two_qubit_gate_product_tabulation(
+            protocols.unitary(base_gate),
+            max_infidelity,
+            sample_scaling=sample_scaling,
+            allow_missed_points=allow_missed_points,
+            random_state=random_state,
+        )
+
+    @property
+    def base_gate(self) -> 'cirq.Gate':
+        """Get the base_gate from initialization."""
+        return self._base_gate
+
+    @property
+    def tabulation(self) -> two_qubit_gate_tabulation.TwoQubitGateTabulation:
+        """Get the GateTabulation object associated with base_gate."""
+        return self._tabulation
+
+    def _decompose_two_qubit_operation(self, op: 'cirq.Operation', _) -> 'cirq.OP_TREE':
+        if not protocols.has_unitary(op):
+            return NotImplemented
+
+        if protocols.has_kraus(op):
+            e_fid = measures.entanglement_fidelity(cast(protocols.SupportsKraus, op))
+            if e_fid > 1.0 - self._tabulation.max_expected_infidelity:
+                return []  # we are close enough to identity.
+
+        q0, q1 = op.qubits
+        decomp = self._tabulation.compile_two_qubit_gate(protocols.unitary(op))
+        ret = []
+        for i in range(len(decomp.local_unitaries) - 1):
+            mats = decomp.local_unitaries[i]
+            for mat, q in zip(mats, [q0, q1]):
+                phxz_gate = single_qubit_decompositions.single_qubit_matrix_to_phxz(mat)
+                if phxz_gate is not None:
+                    ret.append(phxz_gate(q))
+            ret.append(self._base_gate(q0, q1))
+
+        mats = decomp.local_unitaries[-1]
+        for mat, q in zip(mats, [q0, q1]):
+            phxz_gate = single_qubit_decompositions.single_qubit_matrix_to_phxz(mat)
+            if phxz_gate is not None:
+                ret.append(phxz_gate(q))
```

Comment:
I think the right level of abstraction for this logic of 

1. Creating a tabulation in the` __init__` function and
2. Using the created tabulation for decomposing any given two qubit operation `op` into gates from the target gateset

should reside as a standalone class in the `cirq-core/cirq/transformers/heuristic_decompositions/two_qubit_gate_tabulation.py` OR as a standalone transformer in `cirq-core/cirq/transformers`, decorated with @cirq.transformer (which is also requested in https://github.com/quantumlib/Cirq/issues/4059)

The main reason behind this proposal is that we'd like to reuse this logic at different places, like individual transformers and different target gatesets that wish to "fall-back" on the approximate decomposition the given operation to be decomposed is not part of a known set. (eg: the `SycamoreTargetGateset` introduced in https://github.com/quantumlib/Cirq/pull/5054).